requests/adapters.py

```diff
@@ -388,7 +388,10 @@ def send(self, request, stream=False, timeout=None, verify=True, cert=None, prox
                     low_conn.endheaders()
 
                     for i in request.body:
-                        low_conn.send(hex(len(i))[2:].encode('utf-8'))
+                        chunk_size = len(i)
+                        if chunk_size == 0:
+                            break
```

Comment:
If we're trying to be consistent with existing behaviour, I think we're setting the bar too low. With the PR open against 3.0.0 we can design a sensible behaviour, rather than a compatible one, and I think that the sensible behaviour is to do our best to consume the entire iterator.